[#422] [Chore] [Wiki] Update the Fastlane page (#424)
nimblehq/ios-templates: .github/wiki/Fastlane.md

> Fastlane is a part of the automation tool for the development and release process of the project. By converting to Fastlane.swift, we can now write our configuration using Xcode in Swift - the language we know and love from the world of iOS development. This document will mention some crucial parts of the pipeline. 

Reviewer: ```suggestion
Fastlane is a part of the automation tool for the development and release process of the project. By converting to Fastlane.swift, we can now write our configuration using Xcode in Swift - the language we know and love from the world of iOS development. This document will mention some crucial parts of the pipeline.
```

<img src="assets/images/infrastructure/fastlane/fastlane.png" width=300>

## Fastfile

### Build and test

> The lane `buildAndTestLane` helps build and test the application with the configuration of the test scheme and test target defined in the `Constant.swift` file.

Reviewer: FYI, we can safely remove the suffix `Lane` for all lanes :D.
```suggestion
The lane `buildAndTest` helps build and test the application with the configuration of the test scheme and test target defined in the `Constant.swift` file.
```

Reviewer: Feel free to remove all the rest lanes, which still contain the suffix `Lane` in both the description and executing command🙏  Thank youu.

Example:

```

> $ bundle exec fastlane buildAndTestLane

Reviewer: same here when executing the lane.
```suggestion
$ bundle exec fastlane buildAndTest
```

```

> After running this lane, it can generate the code coverage report.

Reviewer 2: ```suggestion
After running this lane successfully, it will generate the code coverage report.
```

> See more:
> 
> - [Constant.swift](#constantswift)

### Synchronize the provisioning profile

As the developer team has many developers, it is recommended to synchronize the certificates and profiles across the team during the development process.

The `fastlane match` is here for sharing one code signing identity across the development team to simplify the codesigning setup and prevent code signing issues.

> To synchronize the machine with the certificate and profiles, which are stored in a match repository, please use the following lanes:

Reviewer 2: ```suggestion
To synchronize the machine with the certificates and profiles, which are stored in a match repository, please use the following lanes:
```

| `syncDevelopmentCodeSigningLane` | `syncAdHocStagingCodeSigningLane` | `syncAdHocProductionCodeSigningLane` | `syncAppStoreCodeSigningLane` |
|---|---|---|---|
| Synchronize the Development match signing for the Staging build. | Synchronize the Ad Hoc match signing for the Staging build. | Synchronize the Ad Hoc match signing for the Production build. | Synchronize the App Store match signing for the Production build. |

Example:

```

> $ bundle exec fastlane syncDevelopmentCodeSigningLane
> $ bundle exec fastlane syncAdHocStagingCodeSigningLane
> $ bundle exec fastlane syncAdHocProductionCodeSigningLane
> $ bundle exec fastlane syncAppStoreCodeSigningLane

Reviewer: same here 🙏 
```suggestion
$ bundle exec fastlane syncDevelopmentCodeSigning
$ bundle exec fastlane syncAdHocStagingCodeSigning
$ bundle exec fastlane syncAdHocProductionCodeSigning
$ bundle exec fastlane syncAppStoreCodeSigning
```

```

### Register a new device

To register a new device and synchronize the new device across the development team, use the lane `registerNewDevice` and provide the device UDID along with the device name:

Example:

```
$ bundle exec fastlane registerNewDevice
```

### Build and upload the application

> To build and upload the application to distribution tools, like Firebase or App Store Connect, please use these lanes:

Reviewer 2: ```suggestion
To build and upload the application to distribution platforms, like Firebase or App Store Connect, please use these lanes:
```

|---|---|---|---|
| To upload the Staging build and Staging dSYM file to Firebase. | To upload the Production build and Production dSYM file to Firebase. | To upload the Production build to App Store. | To upload the Production build to TestFlight. |

Example: 

```
$ bundle exec fastlane buildStagingAndUploadToFirebase
$ bundle exec fastlane buildProductionAndUploadToFirebase
$ bundle exec fastlane buildAndUploadToAppStore
$ bundle exec fastlane buildAndUploadToTestFlight
```

> See more:
> 
> - [Build.swift](#buildswift)
> - [Distribution.swift](#distributionswift)

## Matchfile

Define the essential information to synchronize the certificates and profiles across the development team.

> See more:
> 
> - [Match.swift](#matchswift)
> - [Synchronize the provisioning profile](#synchronize-the-provisioning-profile)

## Constants folder

### Constant.swift

It contains the key/value pairs of constants used during the development and release process.

### Secret.swift

It contains the key/value pairs of environment variables used during the development and release process.

## Managers folder

### Build.swift

> The `Build` helps build and sign the application. There are two main functions:

Reviewer 2: ```suggestion
The `Build` helps building and signing the application. There are two main functions:
```

Author: I will keep the pure verb. Ref: https://www.oxfordlearnersdictionaries.com/definition/american_english/help_1

![Screenshot 2023-01-06 at 13 36 35](https://user-images.githubusercontent.com/24598204/210944516-a043c077-684f-4804-91f4-65f8f8819051.png)

![Screenshot 2023-01-06 at 13 37 09](https://user-images.githubusercontent.com/24598204/210944569-84a2f0cf-14a6-4c2c-b3bd-42b46158c0b2.png)

Reviewer 3: I think you can start this sentence with no `The`.
```suggestion
`Build.swift` helps build and sign the application. There are two main functions:
```

| `adHoc` | `appStore` |
|---|---|
| Build and sign the application with the `adHoc` distribution and the environment parameter (`staging` or `production`) | Build and sign the application with the `appStore` distribution. |

> See more:
>
> - [Build and upload the application](#build-and-upload-the-application)

### Distribution.swift

> The `Distribution` is in charge of distributing the build to the distribution tools, such as Firebase, App Store Connect, and Testflight.

Reviewer 2: ```suggestion
The `Distribution` is in charge of distributing the build to the distribution platforms, such as Firebase, App Store Connect, and Testflight.
```

> The `Distribution` is in charge of distributing the build to the distribution platforms, such as Firebase, App Store Connect, and Testflight.

Reviewer 3: Similarly 
```suggestion
`Distribution.swift` is in charge of distributing the build to the distribution platforms, such as Firebase, App Store Connect, and Testflight.
```

> See more:
>
> - [Build and upload the application](#build-and-upload-the-application)

### Match.swift

The responsibility of `Match` is to synchronize the teams' certificates and profiles during the development process.

> See more:
>
> - [Synchronize the provisioning profile](#synchronize-the-provisioning-profile)
> - [Matchfile](#matchfile)

### Symbol.swift

Technically, the debug Symbol file (dSYM file) is used to de-obfuscate stack traces from crashes on the production app. The dSYM files store the debug symbols for the application. Then services (like Crashlytics) use the dSYM files to replace the symbols in the crash reports with the originating locations in the source code. Hence, the crash reports will be more readable and understandable.

> The `SymbolManager` helps process and upload dSYM file to Firebase. There is only one primary function:

Reviewer 2: ```suggestion
The `SymbolManager` helps processing and uploading dSYM file to Firebase. There is only one primary function:
```

Author: Same here

> The `SymbolManager` helps process and upload dSYM file to Firebase. There is only one primary function:

Reviewer 4: ```suggestion
The `Symbol ` helps processing and uploading dSYM files to Firebase. There is only one primary function:
```

| `uploadToCrashlytics` |
|---|

> | Directly upload the built dSYM file to Crashlytics with the environment parameter  (`staging` or `production`). |

Reviewer: PoW looks correct, but in this `.md`, I think maybe we have one extra space. Please help me to recheck it. Thank you 🙏 .
```suggestion
| Directly upload the built dSYM file to Crashlytics with the environment parameter (`staging` or `production`). |
```
